File: src/lib/autoRoute.ts
// Auto-routing client. The planner lives on the chart server (Go:
// render/autoroute.go) because that is where the ENC depth data is; the
// browser hands it two points and gets back a waypoint list.
//
// The safe depth is the hard constraint — the route never crosses water
// charted shoaler than it — and the ideal depth is the soft one: among safe
// routes, prefer the one that stays that deep. Leave either blank and the
// server uses the boat's configured draft.

import type { LatLng } from "./simplify";

export interface AutoRouteRequest {
  start: LatLng;
  end: LatLng;
  /** Hard minimum depth in feet. Omitted = the module's configured draft. */
  safeDepthFt?: number;
  /** Preferred depth in feet. Omitted = the module's ideal_depth, else 2x draft. */
  idealDepthFt?: number;
  /** No-go buffer off land and shoals, metres. */
  clearanceM?: number;
  /** Charted area classes to steer around. Extensible server-side. */
  avoid?: "restricted"[];
  maxWaypoints?: number;
}

export interface AutoRouteResult {
  waypoints: LatLng[];
  distance_meters: number;
  direct_meters: number;
  /** Shoalest charted depth on the route; null when none of it was charted. */
  min_depth_meters: number | null;
  crossed_unknown: boolean;
  safe_depth_meters: number;
  ideal_depth_meters: number;
  snapped_start: boolean;
  snapped_end: boolean;
  cell_size_meters: number;
  /**
   * How many independently planned runs the route was split into. More than 1
   * means it was too long to plan whole at a useful resolution, and
   * cell_size_meters is the coarsest any section used.
   */
  sections: number;
  warnings?: string[];
}

const METRES_TO_FEET = 3.28084;
const NM = 1852;

export function metresToFeet(m: number): number {
  return m * METRES_TO_FEET;
}

export function metresToNm(m: number): number {
  return m / NM;
}

/** Builds the /noaa-enc/autoroute query. Exported for testing. */
export function autoRouteUrl(base: string, req: AutoRouteRequest): string {
  const p = new URLSearchParams({
    startLat: String(req.start.lat),
    startLon: String(req.start.lng),
    endLat: String(req.end.lat),
    endLon: String(req.end.lng),
  });
  // Only send what the operator actually chose — an omitted parameter means
  // "use the boat's configured value", which is not the same as sending 0.
  if (req.safeDepthFt != null && req.safeDepthFt > 0) p.set("sd", String(req.safeDepthFt));
  if (req.idealDepthFt != null && req.idealDepthFt > 0) p.set("ideal", String(req.idealDepthFt));
  if (req.clearanceM != null && req.clearanceM >= 0) p.set("clearance", String(req.clearanceM));
  if (req.maxWaypoints != null && req.maxWaypoints > 1)
    p.set("max_waypoints", String(req.maxWaypoints));
  if (req.avoid?.length) p.set("avoid", req.avoid.join(","));
  return `${base}/noaa-enc/autoroute?${p.toString()}`;
}

// The chart server may be a different origin from the app (a split
// deployment); /app-config says which. Resolved once and cached — it can't
// change without a page reload.
let tileBasePromise: Promise<string> | null = null;
const HOSTED_TILE_FALLBACK = "https://nycmaps.checkmatemaps.com";

export function resolveTileBase(): Promise<string> {
  if (!tileBasePromise) {
    tileBasePromise = (async () => {
      try {
        const resp = await fetch("/app-config");
        if (!resp.ok) return HOSTED_TILE_FALLBACK;
        const cfg = await resp.json();
        if (cfg && typeof cfg.tileServerBaseURL === "string") {
          return cfg.tileServerBaseURL.replace(/\/$/, "");
        }
      } catch {
        return HOSTED_TILE_FALLBACK;
      }
      return "";
    })();
  }
  return tileBasePromise;
}

export interface OptimizeRequest {
  waypoints: LatLng[];
  safeDepthFt?: number;
  idealDepthFt?: number;
  clearanceM?: number;
  avoid?: "restricted"[];
  maxWaypoints?: number;
  /**
   * Keep every waypoint the operator placed, re-planning only the water
   * between them (the default). False lets the smoother straighten through
   * them and drop the redundant ones.
   */
  keepWaypoints?: boolean;
}

/**
 * Re-plans an existing waypoint list around land, shoals and obstructions.
 * `direct_meters` in the result is the ORIGINAL route's length, so the caller
 * can show what the optimisation cost or saved.
 */
export async function optimizeRoute(req: OptimizeRequest): Promise<AutoRouteResult> {
  const base = await resolveTileBase();
  const resp = await fetch(`${base}/noaa-enc/optimize`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      waypoints: req.waypoints.map((w) => ({ lat: w.lat, lng: w.lng })),
      // Only send what was actually chosen — an omitted field means "use the
      // boat's configured value", which is not the same as sending 0.
      ...(req.safeDepthFt != null && req.safeDepthFt > 0 ? { safe_depth_ft: req.safeDepthFt } : {}),
      ...(req.idealDepthFt != null && req.idealDepthFt > 0
        ? { ideal_depth_ft: req.idealDepthFt }
        : {}),
      ...(req.clearanceM != null && req.clearanceM >= 0 ? { clearance_m: req.clearanceM } : {}),
      ...(req.maxWaypoints != null && req.maxWaypoints > 1
        ? { max_waypoints: req.maxWaypoints }
        : {}),
      ...(req.keepWaypoints != null ? { keep_waypoints: req.keepWaypoints } : {}),
      ...(req.avoid?.length ? { avoid: req.avoid } : {}),
    }),
  });
  if (!resp.ok) {
    let msg = `optimize failed (${resp.status})`;
    try {
      const body = await resp.json();
      if (body && typeof body.error === "string") msg = body.error;
    } catch {
      // non-JSON error body; keep the status message
    }
    throw new Error(msg);
  }
  return (await resp.json()) as AutoRouteResult;
}

/**
 * Plans a route between two points. Throws with the server's own message on
 * failure — "no safe route found …" is the common one and is worth showing
 * verbatim, since it names the depth that made it impossible.
 */
export async function planAutoRoute(req: AutoRouteRequest): Promise<AutoRouteResult> {
  const base = await resolveTileBase();
  const resp = await fetch(autoRouteUrl(base, req));
  if (!resp.ok) {
    let msg = `auto-route failed (${resp.status})`;
    try {
      const body = await resp.json();
      if (body && typeof body.error === "string") msg = body.error;
    } catch {
      // non-JSON error body; keep the status message
    }
    throw new Error(msg);
  }
  return (await resp.json()) as AutoRouteResult;
}

/**
 * The human-readable cautions to show beside a planned route: the server's own
 * warnings plus the two derived from the result that a skipper should see
 * before loading it.
 */
export function routeCautions(res: AutoRouteResult): string[] {
  const out = [...(res.warnings ?? [])];
  if (res.crossed_unknown) {
    out.push("part of this route crosses water with no charted depth");
  }
  if (res.sections > 1) {
    out.push(
      `too long to plan in one piece — split into ${res.sections} sections, coarsest grid ${Math.round(res.cell_size_meters)} m`
    );
  }
  if (res.min_depth_meters != null && Number.isFinite(res.min_depth_meters)) {
    out.push(
      `shoalest charted depth on the route: ${metresToFeet(res.min_depth_meters).toFixed(1)} ft`
    );
  }
  return out;
}

File: mobile/lib/routes/auto_route.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:latlong2/latlong.dart';

/// Auto-routing client, ported from src/lib/autoRoute.ts (tests translated
/// alongside). The planner lives on the chart server (Go: render/autoroute.go)
/// because that is where the ENC depth data is; the phone hands it two points
/// and gets back a waypoint list.
///
/// The safe depth is the hard constraint — the route never crosses water
/// charted shoaler than it — and the ideal depth is the soft one: among safe
/// routes, prefer the one that stays that deep. Leave either null and the
/// server uses the boat's configured draft.

const double _metresPerNm = 1852;
const double _metresToFeet = 3.28084;

double metresToFeet(double m) => m * _metresToFeet;
double metresToNm(double m) => m / _metresPerNm;

class AutoRouteResult {
  const AutoRouteResult({
    required this.waypoints,
    required this.distanceMeters,
    required this.directMeters,
    required this.minDepthMeters,
    required this.crossedUnknown,
    required this.safeDepthMeters,
    required this.idealDepthMeters,
    required this.snappedStart,
    required this.snappedEnd,
    required this.cellSizeMeters,
    required this.sections,
    this.warnings = const [],
  });

  final List<LatLng> waypoints;
  final double distanceMeters;
  final double directMeters;

  /// Shoalest charted depth on the route; null when none of it was charted.
  final double? minDepthMeters;
  final bool crossedUnknown;
  final double safeDepthMeters;
  final double idealDepthMeters;
  final bool snappedStart;
  final bool snappedEnd;
  final double cellSizeMeters;

  /// How many independently planned runs the route was split into. More than
  /// 1 means it was too long to plan whole at a useful resolution, and
  /// [cellSizeMeters] is the coarsest any section used.
  final int sections;
  final List<String> warnings;

  static AutoRouteResult fromJson(Map<String, dynamic> j) => AutoRouteResult(
        waypoints: [
          if (j['waypoints'] is List)
            for (final w in j['waypoints'] as List)
              if (w is Map && w['lat'] is num && w['lng'] is num)
                LatLng((w['lat'] as num).toDouble(),
                    (w['lng'] as num).toDouble()),
        ],
        distanceMeters: (j['distance_meters'] as num?)?.toDouble() ?? 0,
        directMeters: (j['direct_meters'] as num?)?.toDouble() ?? 0,
        minDepthMeters: (j['min_depth_meters'] as num?)?.toDouble(),
        crossedUnknown: j['crossed_unknown'] == true,
        safeDepthMeters: (j['safe_depth_meters'] as num?)?.toDouble() ?? 0,
        idealDepthMeters: (j['ideal_depth_meters'] as num?)?.toDouble() ?? 0,
        snappedStart: j['snapped_start'] == true,
        snappedEnd: j['snapped_end'] == true,
        cellSizeMeters: (j['cell_size_meters'] as num?)?.toDouble() ?? 0,
        sections: (j['sections'] as num?)?.toInt() ?? 1,
        warnings: [
          if (j['warnings'] is List)
            for (final w in j['warnings'] as List) w.toString(),
        ],
      );
}

/// Builds the /noaa-enc/autoroute query. Exposed for testing. Only sends
/// what the operator actually chose — an omitted parameter means "use the
/// boat's configured value", which is not the same as sending 0.
String autoRouteUrl(
  String base, {
  required LatLng start,
  required LatLng end,
  double? safeDepthFt,
  double? idealDepthFt,
  double? clearanceM,
  List<String> avoid = const [],
  int? maxWaypoints,
}) {
  final p = <String, String>{
    'startLat': '${start.latitude}',
    'startLon': '${start.longitude}',
    'endLat': '${end.latitude}',
    'endLon': '${end.longitude}',
  };
  if (safeDepthFt != null && safeDepthFt > 0) p['sd'] = '$safeDepthFt';
  if (idealDepthFt != null && idealDepthFt > 0) p['ideal'] = '$idealDepthFt';
  if (clearanceM != null && clearanceM >= 0) p['clearance'] = '$clearanceM';
  if (maxWaypoints != null && maxWaypoints > 1) {
    p['max_waypoints'] = '$maxWaypoints';
  }
  if (avoid.isNotEmpty) p['avoid'] = avoid.join(',');
  return '$base/noaa-enc/autoroute?${Uri(queryParameters: p).query}';
}

/// Plans a route between two points. Throws with the server's own message on
/// failure — "no safe route found …" is the common one and is worth showing
/// verbatim, since it names the depth that made it impossible.
Future<AutoRouteResult> planAutoRoute(
  String base, {
  required LatLng start,
  required LatLng end,
  double? safeDepthFt,
  double? idealDepthFt,
  double? clearanceM,
  List<String> avoid = const [],
  int? maxWaypoints,
  http.Client? client,
}) async {
  final url = Uri.parse(autoRouteUrl(
    base,
    start: start,
    end: end,
    safeDepthFt: safeDepthFt,
    idealDepthFt: idealDepthFt,
    clearanceM: clearanceM,
    avoid: avoid,
    maxWaypoints: maxWaypoints,
  ));
  final resp = await (client?.get(url) ?? http.get(url));
  if (resp.statusCode != 200) {
    var msg = 'auto-route failed (${resp.statusCode})';
    try {
      final body = jsonDecode(resp.body);
      if (body is Map && body['error'] is String) msg = body['error'] as String;
    } catch (_) {
      // non-JSON error body; keep the status message
    }
    throw Exception(msg);
  }
  final body = jsonDecode(resp.body);
  if (body is! Map<String, dynamic>) {
    throw Exception('auto-route returned an unexpected response');
  }
  return AutoRouteResult.fromJson(body);
}

/// The human-readable cautions to show beside a planned route: the server's
/// own warnings plus the ones derived from the result that a skipper should
/// see before loading it.
List<String> routeCautions(AutoRouteResult res) {
  final out = [...res.warnings];
  if (res.crossedUnknown) {
    out.add('part of this route crosses water with no charted depth');
  }
  if (res.sections > 1) {
    out.add('too long to plan in one piece — split into ${res.sections} '
        'sections, coarsest grid ${res.cellSizeMeters.round()} m');
  }
  final minDepth = res.minDepthMeters;
  if (minDepth != null && minDepth.isFinite) {
    out.add('shoalest charted depth on the route: '
        '${metresToFeet(minDepth).toStringAsFixed(1)} ft');
  }
  return out;
}
